connection.go

```diff
@@ -28,18 +28,21 @@ type mysqlContext interface {
 }
 
 type mysqlConn struct {
-	buf              buffer
-	netConn          net.Conn
-	affectedRows     uint64
-	insertId         uint64
-	cfg              *Config
-	maxAllowedPacket int
-	maxWriteSize     int
-	writeTimeout     time.Duration
-	flags            clientFlag
-	status           statusFlag
-	sequence         uint8
-	parseTime        bool
+	buf                 buffer
+	netConn             net.Conn
+	reader              packetReader
+	writer              io.Writer
```

Comment:
Do we need all four (`buf`, `netConn` as well as `reader` and `writer`) or could we just set `buf` / `netConn` as the  `reader` / `writer` (respectively use it to init the compression reader / writer) once during the init phase and then use the abstracted `reader` / `writer`?

I feel like `mysqlConn` is slowly turning into a god object.

Reply:
The reason I did not do that originally is because there are various times in the code base where buf is used to do things other than `readNext` -- e.g. set timeout values and control buffer size manually. It seemed cleaner to have `buf` available directly for these manipulations and a separate abstraction for `readNext` only, to avoid having to access these operations through the `reader`. Let me know what you think!

Reply:
I also noticed that we need to access buf to do  `mc.buf.nc = tlsConn` in packets.go. This seems like it will be very difficult to access if buf becomes hidden in the reader abstraction that is sometimes compressed and sometimes not. Let me know if what I'm saying makes sense, and if you have any ideas about how I can best organize buf and reader together!